File: finanze/domain/backup.py
from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import UUID

from pydantic.dataclasses import dataclass

from domain.cloud_auth import CloudAuthData

CURRENT_PROTOCOL_VERSION = 1


class BackupFileType(str, Enum):
    DATA = "DATA"
    CONFIG = "CONFIG"


class BackupMode(str, Enum):
    OFF = "OFF"
    MANUAL = "MANUAL"
    AUTO = "AUTO"


@dataclass
class BackupInfo:
    id: UUID
    protocol: int
    date: datetime
    type: BackupFileType
    size: int


@dataclass
class BackupsInfo:
    pieces: dict[BackupFileType, BackupInfo]


@dataclass
class BackupsInfoRequest:
    only_local: bool = False


class SyncStatus(str, Enum):
    SYNC = "SYNC"
    PENDING = "PENDING"
    OUTDATED = "OUTDATED"
    MISSING = "MISSING"
    CONFLICT = "CONFLICT"


@dataclass
class FullBackupInfo:
    local: Optional[BackupInfo]
    remote: Optional[BackupInfo]
    last_update: datetime
    has_local_changes: bool
    status: Optional[SyncStatus]


@dataclass
class FullBackupsInfo:
    pieces: dict[BackupFileType, FullBackupInfo]


@dataclass
class BackupProcessRequest:
    protocol: int
    password: str
    payload: bytes


@dataclass
class BackupProcessResult:
    payload: bytes


@dataclass
class BackupTransferPiece:
    id: UUID
    protocol: int
    date: datetime
    type: BackupFileType
    payload: bytes


@dataclass
class BackupPieces:
    pieces: list[BackupTransferPiece]


@dataclass
class BackupUploadParams:
    pieces: BackupPieces
    auth: CloudAuthData


@dataclass
class BackupDownloadParams:
    types: list[BackupFileType]
    auth: CloudAuthData


@dataclass
class BackupInfoParams:
    auth: CloudAuthData


@dataclass
class UploadBackupRequest:
    types: list[BackupFileType]
    force: bool = False


@dataclass
class ImportBackupRequest:
    types: list[BackupFileType]
    password: Optional[str]
    force: bool = False


@dataclass
class BackupSyncResult:
    pieces: dict[BackupFileType, FullBackupInfo]


@dataclass
class BackupSettings:
    mode: BackupMode

File: mobile/src/domain/core/backup.ts
import { CloudAuthData } from "./cloudAuth"

export enum BackupFileType {
  DATA = "DATA",
  CONFIG = "CONFIG",
}

export enum BackupMode {
  OFF = "OFF",
  MANUAL = "MANUAL",
  AUTO = "AUTO",
}

export interface BackupInfo {
  id: string
  protocol: number
  date: string
  type: BackupFileType
  size: number
}

export interface BackupsInfo {
  pieces: Partial<Record<BackupFileType, BackupInfo>>
}

export interface BackupsInfoRequest {
  onlyLocal?: boolean
}

export enum SyncStatus {
  SYNC = "SYNC",
  PENDING = "PENDING",
  OUTDATED = "OUTDATED",
  MISSING = "MISSING",
  CONFLICT = "CONFLICT",
}

export interface FullBackupInfo {
  local: BackupInfo | null
  remote: BackupInfo | null
  lastUpdate: string
  hasLocalChanges: boolean
  status: SyncStatus | null
}

export interface FullBackupsInfo {
  pieces: Partial<Record<BackupFileType, FullBackupInfo>>
}

export interface BackupProcessRequest {
  protocol: number
  password: string
  payload: Uint8Array
}

export interface BackupProcessResult {
  payload: Uint8Array
}

export interface BackupTransferPiece {
  id: string
  protocol: number
  date: string
  type: BackupFileType
  payload: Uint8Array
}

export interface BackupPieces {
  pieces: BackupTransferPiece[]
}

export interface BackupUploadParams {
  pieces: BackupPieces
  auth: CloudAuthData
}

export interface BackupDownloadParams {
  types: BackupFileType[]
  auth: CloudAuthData
}

export interface BackupInfoParams {
  auth: CloudAuthData
}

export interface UploadBackupRequest {
  types: BackupFileType[]
  force?: boolean
}

export interface ImportBackupRequest {
  types: BackupFileType[]
  password: string | null
  force?: boolean
}

export interface BackupSyncResult {
  pieces: Partial<Record<BackupFileType, FullBackupInfo>>
}

export interface BackupSettings {
  mode: BackupMode
}
